mobilecoind-dev-faucet/src/slam/mod.rs

```diff
@@ -0,0 +1,535 @@
+                        counter,
+                        utxo_record,
+                        params,
+                        mobilecoind_api_client,
+                        logger,
+                    )
+                    .await
+                    {
+                        Ok(prepared_utxo) => {
+                            prepared_utxos_sender
+                                .send(prepared_utxo)
+                                .await
+                                .map_err(|err| {
+                                    format!("Unexpected error when preparing: {}", err)
+                                })?;
+                            self.num_prepared_utxos.fetch_add(1, Ordering::SeqCst);
+                        }
+                        Err(err) => {
+                            log::error!(
+                                logger,
+                                "Could not prepare utxo #{}, skipping: {}",
+                                counter,
+                                err
+                            );
+                        }
+                    }
+                }
+                Err(GetUtxoError::Busy) => {
+                    // Try again later after the worker has had a chance to split more utxos
+                    tokio::time::sleep(worker.get_worker_poll_period()).await;
+                }
+                Err(err) => {
+                    // Any more serious error, like "funds depleted", means we cannot complete the
+                    // slam
+                    return Err(format!("Cannot obtain more utxos, aborting slam: {}", err));
+                }
+            }
+        }
+
+        // Close this so that workers that get to the end of the queue won't block
+        prepared_utxos_sender.close();
+
+        // Get the network state, which contains block height, now so that it is
+        // fresh for the slam workers
+        let network_state = {
+            let resp = mobilecoind_api_client
+                .get_network_status(&Default::default())
+                .map_err(|err| format!("Failed getting network status: {}", err))?;
+            self.block_height
+                .store(resp.network_highest_block_index, Ordering::SeqCst);
+            resp
+        };
+
+        // Now, we spawn worker threads, which build txs and
+        // submits them, with retries, in parallel
+        self.phase
+            .store(SlamState::SUBMITTING_TRANSACTIONS, Ordering::SeqCst);
+        log::info!(logger, "Slam status: {}", self.get_status().unwrap());
+        let begin_submit_time = Instant::now();
+        let prepare_time = begin_submit_time.duration_since(begin_prepare_time);
+
+        // Note: Workers are std::thread rather than tokio::task because mc_connection
+        // currently only has blocking APIs If we get an async-api for
+        // attestation and tx submission, we could change this to be tasks and just use
+        // the tokio thread pool. maybe some advantages?
```

Comment:
Probably biggest advantage it would bring is it could use its underlying green threading to split up i/o work properly on the threads.

It would be something potentially worth measuring if this way of "slamming" ends up being slower than we'd like it to be.

Reply:
i have no doubt it's more efficient in an absolute sense, but I'm not sure it's going to be a measureable difference in this problem setting.

So like, if green threading enables us to avoid a bunch of OS level thread context switches, and the benefit is, we save a bunch of 5 microseconds context switches, and maybe several of these per submission... it's still like, the blockchain is doing at most 50 tx / s. So saturating the blockchain a second faster, or even a few seconds faster, I'm not sure it really matters -- as long as we saturate the submission queues and hit peak so that we can measure the peak then I think it's fine

Reply:
Probably doesn't matter. I think this approach to async is good - get fancy only when we need to get fancy.